File: scripts/pahma/arks/checknos.py
import sys
import re
import urllib.parse

for musnos in sys.stdin.readlines():

    (sortNum, original) = musnos.strip().split('\t')
    match = re.match(r'^([0-9a-zA-Z]+)-(.*)', original)
    if match is None:
        #ARK = urllib.parse.quote('x' + original).replace('-','%2D').replace('.','%2E')
        ARK = urllib.parse.quote('x' + original).replace('-','%2D').replace('.','@')
    else:
        left, right = match.group(1), match.group(2)
        #ARK = urllib.parse.quote(left.zfill(3) + right.zfill(7)).replace('-','%2D').replace('.','%2E')
        ARK = urllib.parse.quote('1' + left.zfill(2) + right.zfill(7)).replace('-','%2D').replace('.','@')

    # reverse the "arkification"
    unARK = urllib.parse.unquote(ARK).replace('@','.')
    if unARK[0] == 'x':
        unARK = unARK[1:]
    else:
        left, right = unARK[1:3], unARK[3:]
        left = left.lstrip('0')
        right = right.lstrip('0')
        unARK = left + '-' + right

    # check if the reverse worked...
    reversible = '' if original == unARK else 'not reversible'

    print(f'%s\t%s\t%s\t%s' % (original, ARK, unARK, reversible))

File: scripts/pahma/arks/checknos.pl
use strict;

my %chars;
my %patterns;

while (<>) {
    # parse input line, clean up a bit
    chomp;
    #s/ +$//;
    #s/ +\t/\t/;
    my ($sortNum, $original) = split /\t/;
    $_ = $original;

    # make pattern from input
    s/\d+/9/g;
    s/[A-Z]+/A/g;
    s/[a-z]+/a/g;
    my $pattern = $_;
    $patterns{$pattern}++;

    my @chars = split '', $original;
    foreach my $c (@chars) {
        $chars{$c}++;
    }

    # generate ARK suffix
    my $ARK;
    my ($ARK, $left, $right, $extra);

    # keep periods by encoding them as x's
    my $o2 = $original;
    $o2 =~ s/\./X/g;

    # make letters upper case. why not? (strictly speaking not reversible, but we need to avoid lowercase L, etc.)
    $o2 =~ tr/a-z/A-Z/;

    # basic pattern
    my ($left, $right, $extra1, $extra2) = $o2 =~ /^([\d\w]+)\-(\d+)(\w*)\b(.*)$/;
    my $extra = $extra1 . $extra2;
    #$extra =~ s/^[\s\_\-\.\;\:\,\+\/\#]+//; # remove initial non-filing chars (i.e. punctuation)
    #                                        # NB: this may be problematic in some cases.

    if ($left && $right) {
        $ARK = sprintf "%2s%6s", ($left, $right, $extra);
        $ARK =~ s/ /0/g;
        $ARK .= $extra;
    }
    else { # fallback : see prose above
        $ARK = 'noARK';
    }

    # reverse the "arkification"
    my $unARK = $ARK;
    $unARK =~ s/X/./g;
    my ($left, $right, $extra) = $unARK =~ /^(.{2})(.{6})(.*)/;
    $left =~ s/^0+//;
    $right =~ s/^0+//;
    $unARK = $left . '-' . $right . $extra;

    # check if the reverse worked...
    my $uc_original = $original;
    $uc_original =~ tr/a-z/A-Z/;
    my $reversible = $unARK eq $uc_original ? '' : 'not reversible';

    #print "$left x $right x $extra\n";
    #print "$unARK eq $ARK\n";
    printf "%s\t%s\t%s\t%s\t%s\t%s\t%s\n", ($ARK, $original, $pattern, $sortNum, $unARK, $uc_original, $reversible);
}

print STDERR "\ncharacters\n";
foreach my $char (sort { $chars{$b} <=> $chars{$a} } keys %chars) {
    printf STDERR "%s\t%s\n", ($char, $chars{$char})
}

print STDERR "\npatterns\n";
foreach my $pattern (sort { $patterns{$b} <=> $patterns{$a} } keys %patterns) {
    printf STDERR "%s\t%s\n", ($pattern, $patterns{$pattern})
}